cpp/tensorrt_llm/kernels/cutlass_kernels/moe_gemm/moe_kernels.cu

```diff
@@ -3692,15 +3854,22 @@ void CutlassMoeFCRunner<T, WeightType, OutputType, InputType, BackBoneType, Enab
         if constexpr (!use_w4afp8)
         {
             gemm1_input = applyPrequantScale(smoothed_act_, permuted_data_, quant_params.groupwise.fc1.act_scales,
-                num_valid_tokens_ptr, expanded_num_rows, hidden_size, use_awq, stream);
+                num_valid_tokens_ptr, expanded_num_rows, hidden_size, use_awq, stream, quant_params);
         }
         sync_check_cuda_error(stream);
-        Self::gemm1(moe_gemm_runner_, blockscale_gemm_runner, gemm1_input, fc1_result_, glu_inter_result_,
+
+        // Opportunistically apply FC2 prequant scaling in FC1 doActivation kernel if applicable
+        bool const fuse_fc2_prequant_scale = use_awq && is_gated_activation;
+        void const* fc2_prequant_scale_ptr = fuse_fc2_prequant_scale ? quant_params.groupwise.fc2.act_scales : nullptr;
+        // Match the FC2 act buffer bound to respective TMA desc defined in setupTmaWarpSpecializedInputs()
+        T* gemm1_output = fuse_fc2_prequant_scale ? reinterpret_cast<T*>(smoothed_act_) : fc1_result_;
+        Self::gemm1(moe_gemm_runner_, blockscale_gemm_runner, gemm1_input, gemm1_output, glu_inter_result_,
             expert_first_token_offset_, gemm1_tma_ws_input, fc1_expert_weights, fc1_expert_biases, num_valid_tokens_ptr,
             fc1_int_scales, fc1_fp8_dequant, use_wfp4afp8 ? fc2_wfp4afp8_quant_scale : fc2_fp8_quant,
             fc1_fp4_act_scale_, fc2_fp4_act_scale_, quant_params, num_rows, expanded_num_rows,
             expected_tokens_per_expert, hidden_size, inter_size, num_experts_per_node, fc1_activation_type,
-            alpha_scale_ptr_array_fc1_, !use_lora, stream, *gemm1_config_, false, nullptr, nullptr);
+            alpha_scale_ptr_array_fc1_, !use_lora, stream, *gemm1_config_, false, nullptr, nullptr,
+            fc2_prequant_scale_ptr);
```

Comment:
Does the autotuner need updated to enable this?

Reply:
Good point. I haven't considered autotuner yet because prequant fusion hardcoded as part of W4A8_AWQ recipe hence not tunable. Or do you mean something else?